fish-rust/src/common.rs

```diff
@@ -1985,6 +1969,122 @@ pub fn get_by_sorted_name<T: Named>(name: &wstr, vals: &'static [T]) -> Option<&
     }
 }
 
+/// Takes ownership of a variable and `Deref`s/`DerefMut`s into a projection of that variable.
+///
+/// Can be used as a workaround for the lack of `MutexGuard::map()` to return a `MutexGuard`
+/// exposing only a variable of the Mutex-owned object.
+pub struct Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    value: T,
+    view: F1,
+    view_mut: F2,
+}
+
+impl<T, V, F1, F2> Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    pub fn new(owned: T, project: F1, project_mut: F2) -> Self {
+        Projection {
+            value: owned,
+            view: project,
+            view_mut: project_mut,
+        }
+    }
+}
+
+impl<T, V, F1, F2> Deref for Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        (self.view)(&self.value)
+    }
+}
+
+impl<T, V, F1, F2> DerefMut for Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        (self.view_mut)(&mut self.value)
+    }
+}
+
+/// A trait to make it more convenient to pass ascii/Unicode strings to functions that can take
+/// non-Unicode values.
+///
+/// This is only implemented for owned types where an owned instance will skip allocations (e.g.
+/// `CString` can return `self`) but not implemented for owned instances where a new allocation is
+/// always required (e.g. implemented for `&wstr` but not `WideString`) because you might as well be
+/// left with the original item if we're going to allocate from scratch in all cases.
+pub trait ToCString {
+    fn to_cstring(self) -> CString;
+}
+
+/// Safely converts from `&str` to a `CString`, respecting the user's locale.
+///
+/// We cannot convert directly from a unicode `&str` to a CString as the user's current environment
+/// might not be UTF-8.
+impl ToCString for &str {
+    fn to_cstring(self) -> CString {
+        if self.is_ascii() {
+            CString::new(self).unwrap()
+        } else {
+            // Use wcs2zstring to convert to a CString in the user's current locale.
+            self::wcs2zstring(self)
+        }
+    }
+}
+
+impl ToCString for CString {
+    fn to_cstring(self) -> CString {
+        self
+    }
+}
+
+impl ToCString for &CStr {
+    fn to_cstring(self) -> CString {
+        self.to_owned()
+    }
+}
+
+/// Safely converts from `&wstr` to a `CString`, taking into account non-Unicode values that have
+/// been shifted into the private-use range by using [`wcs2zstring()`].
+impl ToCString for &wstr {
```

Comment:
You cannot do that without giving up all other impls because "upstream may someday add AsRef<wstr> to CString, making it ambiguous which trait impl should be used."